src/Functions/array/arraySum.cpp

```diff
@@ -75,7 +76,8 @@ struct ArraySumImpl
             size_t pos = 0;
             for (size_t i = 0; i < offsets.size(); ++i)
             {
-                res[i] = x * (offsets[i] - pos);
+                /// Just multiply the value by array size.
+                res[i] = Result(x) * (offsets[i] - pos);
```

Comment:
It's non-realistic that overflow here will alter the result.
The cases like

`SELECT arraySum(x -> toInt32(0x80000000), [1, 2])`
and
`SELECT arraySum(x -> toFloat32(1000000.1), [1, 2])`
work correct.